activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,155 @@
+        if returning && !connection.supports_insert_returning?
+          raise ArgumentError, "#{connection.class} does not support :returning"
+        end
+
+        unless %i{ raise skip update }.member?(on_duplicate)
+          raise NotImplementedError, "#{on_duplicate.inspect} is an unknown value for :on_duplicate. Valid values are :raise, :skip, and :update"
+        end
+
+        if on_duplicate == :skip && !connection.supports_insert_on_duplicate_skip?
+          raise ArgumentError, "#{connection.class} does not support skipping duplicates"
+        end
+
+        if on_duplicate == :update && !connection.supports_insert_on_duplicate_update?
+          raise ArgumentError, "#{connection.class} does not support upsert"
+        end
+
+        if unique_by && !connection.supports_insert_conflict_target?
+          raise ArgumentError, "#{connection.class} does not support :unique_by"
+        end
+      end
+
+      def to_sql
+        connection.build_insert_sql(ActiveRecord::InsertAll::Builder.new(self))
+      end
+
+      def keys
+        inserts.present? ? inserts.first.keys.map(&:to_s) : []
+      end
+
+      def readonly_columns
+        primary_keys + model.readonly_attributes.to_a
+      end
+
+      def unique_by_columns
+        unique_by ? unique_by.fetch(:columns).map(&:to_s) : []
+      end
+
+      def primary_keys
+        Array.wrap(model.primary_key)
+      end
+
+
+      class Builder
+        attr_reader :model
+
+        def initialize(insert_all)
+          @insert_all, @model, @connection = insert_all, insert_all.model, insert_all.connection
+        end
+
+        def skip_duplicates?
+          insert_all.on_duplicate == :skip
+        end
+
+        def update_duplicates?
+          insert_all.on_duplicate == :update
+        end
+
+        def into
+          [ "INTO ", model.quoted_table_name, "(", columns_list, ") " ].join
+        end
+
+        def values_list
+          columns = connection.schema_cache.columns_hash(model.table_name)
+
+          values_list = insert_all.inserts.map do |attributes|
+            attributes = attributes.stringify_keys
+
+            unless attributes.keys.to_set == keys.to_set
+              raise ArgumentError, "All objects being inserted must have the same keys"
+            end
+
+            keys.map do |name|
+              column = columns[name]
+              type = connection.lookup_cast_type_from_column(column)
+              bind = Relation::QueryAttribute.new(name, attributes[name], type)
+              connection.with_yaml_fallback(bind.value_for_database)
+            end
+          end
+
+          Arel::InsertManager.new.create_values_list(values_list).to_sql
+        end
+
+        def returning
+          quote_columns(insert_all.returning).join(",") if insert_all.returning
+        end
+
+        def conflict_target
+          return unless conflict_columns
+          sql = [ "(", quote_columns(conflict_columns).join(","), ")" ]
```

Comment:
🤔I made the other two `Array#join` -> String Interpolation changes. In this case, is the `Array#join` more awkward than unfreezing the string to modify it?

I also _slightly_ prefer `Array#join` when the interpolated part would contain another string (the `","` passed to `.join` — which is a string within a string)

Reply:
Another suggestion (using `','` instead of `","`):

```suggestion
          sql = +"(#{quote_columns(conflict_columns).join(',')})"
```

We have some other SQL string visitors (Arel, and schema creation), both are also using string concatation.

https://github.com/rails/rails/blob/383efb0138a8e8b69c85d787fae5fc28e214e6a3/activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb#L23-L28

https://github.com/rails/rails/blob/383efb0138a8e8b69c85d787fae5fc28e214e6a3/activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb#L23-L28